[DI - Step 1] 미스터코 미션 제출합니다.

eunsukko opened the pull request:
안녕하세요~ 미션을 제출하게된 미스터코입니다!

아직은 di에 대해서 그림이 안그려져서 일단 생각나는 데로 구현해보았습니다.
slipp 코드의 경우도 `@Service, @Repository` 를 추가하기 위해서 수정해보았습니다.

현재 BeanFactory 의 경우
1. 위상정렬을 통해서 생성하는 순서를 미리 구한다. (구해진 순서는 빈을 생성할 때, 특정 빈에 대해서 생성시 필요한 빈들은 구해진 순서에서 앞에 존재함을 보장합니다)
2. 구해진 순서에 맞게 생성합니다.

감사합니다!

eunsukko committed:
feat: di 기능 구현

eunsukko committed:
refactor: BeanFactory 의미있는 함수, 변수 명으로 변경

eunsukko committed:
refactor: BeanFactory 에서 di 생성순서 부분 분리

eunsukko committed:
refactor: di 모듈

eunsukko committed:
refactor: mvc 프레임워크에 di 프레임워크 적용

eunsukko committed:
refactor: di 모듈 (생성순서 구하는 부분 수정)

hongsii commented on slipp/src/test/java/slipp/dao/UserDaoTest.java:
```diff
@@ -37,7 +38,8 @@ public void crud() throws Exception {
     @Test
     public void findAll() throws Exception {
-        UserDao userDao = UserDao.getInstance();
+//        UserDao userDao = UserDao.getInstance();
```
굳이 주석으로 남겨두신 이유가 있을까요?

eunsukko replied:
테스트에서도 di 프레임워크를 통해서 주입해줘야하지 않을까 라고 생각해서 남겨놓았었습니다..!
(일단 TODO 를 통해서 표시를 해놓았습니다. )

SpringBoot 에서는 `@ExtendWith` 를 통해서 주입을 했었는데.. 아직은 이 부분이 이해가 안되기도 해서 구현을 하진 않았습니다.
'BeanFactory 가 초기화되고 사용되는 부분'이 좀 더 이해가 되면.. 구현을 시도해 볼 수 있을 것 같아요

hongsii commented on nextstep-di/src/test/java/nextstep/di/factory/BeanFactoryTest.java:
```diff
@@ -18,7 +18,7 @@
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
 public class BeanFactoryTest {
```
사이클을 체크하는 기능을 추가하셨으니 이와 관련된 테스트 코드도 추가해보면 어떨까요?

eunsukko replied:
테스트 코드를 추가해보았습니다..! 

현재는 따로 package 를 구성해서 사이클이 존재하는 경우를 생성해놓았습니다.

이외에도
    - 생성에 필요한 빈이 존재하지 않을 경우
    - 특정 인터페이스에 대한 빈이 존재하지 않을 경우
    - 특정 인터페이스에 대한 빈이 여러개 존재하는 경우
에 대한 테스트를 진행해보았습니다. (테스트를 짜다보니.. 조금씩은 BeanFactory 를 사용하는 사용자 입장에서 생각하게 되네요)

일단.. 좋은 방법이 생각이 나지 않아서.. 각각의 경우에 맞는 패키지를 만들고 그 내부에 특정 상황들을 만들어 놓았습니다

hongsii commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactoryUtils.java:
```diff
@@ -48,4 +55,17 @@
+        log.debug("type: {}", type.toString());
+
+        return Optional.ofNullable(getInjectedConstructor(type))
+                .orElse((Constructor) type.getConstructors()[0]);
```
`Optional` 은 null을 반환해 생기는 에러를 방지하기 위해 만들어졌으며, `Optional` 을 생성하는 비용이 싼 편은 아니라 단순히 null 체크하는 용도로는 사용을 지양하는게 좋습니다.

* [Item 12: Avoid Chaining Optional's Methods With the Single Purpose of Getting a Value](https://dzone.com/articles/using-optional-correctly-is-not-optional)

hongsii commented on nextstep-di/src/test/java/nextstep/di/scanner/TypeScannerTest.java:
```diff
@@ -0,0 +1,29 @@
+        assertThat(types.size()).isEqualTo(1);
+        for (Class<?> expectedType : expectedTypes) {
+            assertThat(types).contains(expectedType);
+        }
```
assertj를 활용하면 테스트 코드의 가독성을 높일 수 있습니다.

``` java
assertThat(types)
    .hasSize(1)
    .containsExactlyElementsOf(expectedTypes);
```

eunsukko replied:
감사합니다.!

의식하지 못했었는데... 왠만한 검증은 다 제공해주는 메소드를 통해서 가능할 것 같네요..!

hongsii reviewed (changes requested):
안녕하세요 미스터코! 리뷰가 늦어져서 죄송합니다 ㅠㅠ
전체적으로 잘 구현해주셨네요!
몇 가지 피드백드렸으니 참고해서 반영해주세요 :)

eunsukko committed:
refactor: 주석의 내용 변경

eunsukko committed:
refactor: 테스트코드에서 검증 부분 명확하게 하기 (TypeScannerTest)

eunsukko committed:
refactor: Optional 남용하는 부분 리팩토링 (BeanFactoryUtils)

eunsukko committed:
feat: BeanFactoryTest 보강

- 순환 참조가 존재하는 경우
- 생성에 필요한 빈이 존재하지 않을 경우
- 특정 인터페이스에 대한 빈이 존재하지 않을 경우
- 특정 인터페이스에 대한 빈이 여러개 존재하는 경우

hongsii commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,92 @@
+    }
+
+    private void validateCanBeBean(Class<?> type) {
+        if (!type.isInterface() && !preInstantiatedTypes.contains(type)) {
```
이 로직이 필요한 이유가 있을까요? `preInstantiatedTypes` 를 순회하면서 분기문이 호출되기 때문에 항상 참이 됩니다.

eunsukko replied:
앗.. 답변이 늦었습니다... ㅠ!! 

TopologySort 에서 기존에 존재하지 않는 타입이 추가될 수가 있습니다. (예를들어 preInstantiatedTypes 에는 등록되지 않는 타입들이 `@Inject` 생성자의 파라미터에 있는 경우입니다. 잘못 사용하는 경우이기는 한데.. 이런 경우를 걸러주기 위해서 구현했었습니다)

hongsii commented on nextstep-di/src/main/java/nextstep/di/factory/BeanFactory.java:
```diff
@@ -1,29 +1,92 @@
+    }
+
+    private Object instantiate(Constructor<?> constructor) {
+        return BeanUtils.instantiateClass(constructor, getBeansSatisfiedWith(getParameterTypes(constructor)));
```
인터페이스가 빈으로 등록되려는 경우에도 정상적으로 동작하나요?

hongsii reviewed (approved):
피드백 반영을 잘 해주셨습니다 👍 
추가적으로 피드백 드렸으니, 2단계에서 같이 반영해주세요!
